KAFKA-13919: expose log recovery metrics

showuon opened the pull request:
Implementation for [KIP-831](https://cwiki.apache.org/confluence/display/KAFKA/KIP-831%3A+Add+metric+for+log+recovery+progress). 

In this PR, I did:
1. add `remainingLogsToRecover` metric for the number of remaining logs for each log.dir to be recovered
2. add `remainingSegmentsToRecover` metric for the number of remaining segments for the current log assigned to the recovery thread.
3. remove these metrics after log loaded completely
4. add tests (it's quite complicated to test it...)

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

showuon committed:
add metrics

showuon committed:
add log

showuon committed:
des

showuon committed:
fix

showuon committed:
seg

showuon committed:
++

showuon committed:
update log

showuon committed:
segments

showuon committed:
string

showuon committed:
tostring

showuon committed:
add segments

showuon committed:
check

showuon committed:
statck

showuon committed:
test

showuon committed:
opt

showuon committed:
name

showuon committed:
-1

showuon committed:
delete log

showuon committed:
remove metrics in the end

showuon committed:
Merge branch 'logMetrics' of https://github.com/showuon/kafka into KAFKA-13919

showuon committed:
KAFKA-13919: refactor

showuon committed:
KAFKA-13919: fix remaining log

showuon committed:
Merge branch 'trunk' of https://github.com/apache/kafka into KAFKA-13919

showuon committed:
KAFKA-13919: add tests

showuon commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -282,7 +285,8 @@ class LogManager(logDirs: Seq[File],
       topicId = None,
-      keepPartitionMetadataFile = keepPartitionMetadataFile)
+      keepPartitionMetadataFile = keepPartitionMetadataFile,
+      numRemainingSegments = numRemainingSegments)
```
pass `numRemainingSegments` to update the metrics dynamically

showuon commented:
@junrao @tombentley , please take a look when available. Thank you.

tombentley commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -307,6 +313,22 @@ class LogManager(logDirs: Seq[File],
+
+  // factory class for naming the log recovery threads used in metrics
+  class LogRecoveryThreadFactory(val baseName: String) extends ThreadFactory {
+    val threadsNum = new AtomicInteger(0)
```
Singlular `threadNum`?

tombentley commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -410,12 +437,34 @@ class LogManager(logDirs: Seq[File],
+      newGauge("remainingLogsToRecover", () => numRemainingLogs.get(dir.getAbsolutePath),
+        Map("dir" -> dir.getAbsolutePath))
+      for (i <- 0 until numRecoveryThreadsPerDataDir) {
+        val threadName = s"log-recovery-${dir.getAbsolutePath}-$i"
```
Can we encapsulate this in a method, since it's duplicating the logic in the ThreadFactory?

showuon replied:
Good suggestion. Updated.

tombentley commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -410,12 +437,34 @@ class LogManager(logDirs: Seq[File],
+  private[log] def removeLogRecoveryMetrics(): Unit = {
+    for (dir <- logDirs) {
+      removeMetric("remainingLogsToRecover", Map("dir" -> dir.getAbsolutePath))
+      for (i <- 0 until numRecoveryThreadsPerDataDir) {
```
`numRecoveryThreadsPerDataDir` can be changed, so what happens if it changes after the metrics are added and before they're removed?

showuon replied:
You're right. Passed in the current `numRecoveryThreadsPerDataDir` parameter now to prevent this case.

tombentley reviewed (commented):
Thanks @showuon! My main question here is how we should cope with the mutability on the number of recovery threads per log dir.

showuon commented:
@tombentley , good point. I've never thought about dynamic reconfiguring the config things. I'll update the PR. Thanks.

showuon committed:
Merge branch 'trunk' of https://github.com/apache/kafka into KAFKA-13919

showuon committed:
KAFKA-13919: consider numLogRecoveryThread changed dynamically case and add test

showuon commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -410,12 +442,34 @@ class LogManager(logDirs: Seq[File],
+    }
+  }
+
+  private[log] def removeLogRecoveryMetrics(curNumRecoveryThreadsPerDataDir: Int): Unit = {
+    for (dir <- logDirs) {
+      removeMetric("remainingLogsToRecover", Map("dir" -> dir.getAbsolutePath))
+      for (i <- 0 until curNumRecoveryThreadsPerDataDir) {
+        removeMetric("remainingSegmentsToRecover", Map("dir" -> dir.getAbsolutePath, "threadNum" -> i.toString))
```
pass the `curNumRecoveryThreadsPerDataDir` in so that it won't be affected by the dynamically numRecoveryThreads change.

showuon commented on core/src/test/scala/unit/kafka/log/LogManagerTest.scala:
```diff
@@ -638,6 +641,221 @@ class LogManagerTest {
+  }
+
+  @Test
+  def testLogRecoveryMetricsShouldBeRemovedAfterLogRecovered(): Unit = {
```
Add a test to make sure we can handle recovery thread num change well.

showuon commented:
@tombentley , I've updated the PR in this commit: https://github.com/apache/kafka/pull/12347/commits/1d04b6dc30f525cbbe6d3daf9c7c5e5a331896cb . Please take a look again. Thanks.

junrao commented on core/src/main/scala/kafka/log/LogLoader.scala:
```diff
@@ -77,7 +78,8 @@ class LogLoader(
   leaderEpochCache: Option[LeaderEpochFileCache],
-  producerStateManager: ProducerStateManager
+  producerStateManager: ProducerStateManager,
+  numRemainingSegments: ConcurrentMap[String, Int] = new ConcurrentHashMap[String, Int]
```
Could we add the new param to javadoc?

junrao commented on core/src/main/scala/kafka/log/LogLoader.scala:
```diff
@@ -404,12 +406,18 @@ class LogLoader(
     if (!hadCleanShutdown) {
-      val unflushed = segments.values(recoveryPointCheckpoint, Long.MaxValue).iterator
+      val unflushed = segments.values(recoveryPointCheckpoint, Long.MaxValue)
+      val unflushedSize = unflushed.size
```
To be consistent with numFlushed, this probably should be named numUnflushed?

junrao commented on core/src/main/scala/kafka/log/LogLoader.scala:
```diff
@@ -424,9 +432,11 @@ class LogLoader(
           warn(s"Corruption found in segment ${segment.baseOffset}," +
             s" truncating to offset ${segment.readNextOffset}")
-          removeAndDeleteSegmentsAsync(unflushed.toList)
+          removeAndDeleteSegmentsAsync(unflushedIter.toList)
```
This will iterate unflushedIter. If we don't account for that, remaining segments won't be accurate.

showuon replied:
Good call! Updated.

junrao commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -366,25 +392,30 @@ class LogManager(logDirs: Seq[File],
 
               info(s"Completed load of $log with ${log.numberOfSegments} segments in ${logLoadDurationMs}ms " +
                 s"($currentNumLoaded/${logsToLoad.length} loaded in $logDirAbsolutePath)")
+              println("!!! numRemainingSegments:" + numRemainingSegments + ";" + log)
```
Is this needed?

junrao commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -366,25 +392,30 @@ class LogManager(logDirs: Seq[File],
                 // KafkaStorageException might be thrown, ex: during writing LeaderEpochFileCache
                 // And while converting IOException to KafkaStorageException, we've already handled the exception. So we can ignore it here.
+            } finally {
+              updateNumRemainingLogs(dir.getAbsolutePath, logsToLoad.length - numLogsLoaded.get())
```
Hmm, how do we handle concurrency here? It's possible that one thread completes and calls numLogsLoaded.get(), but before it could call updateNumRemainingLogs(), another thread completes and calls updateNumRemainingLogs(). Then this thread continue to call updateNumRemainingLogs(), which updates numRemainingLogs with an outdated count.

showuon replied:
Good catch! Updated.

junrao commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -90,6 +90,10 @@ class LogManager(logDirs: Seq[File],
+  // log dir path -> number of Remaining logs map for remainingLogsToRecover metric
+  private val numRemainingLogs: ConcurrentMap[String, Int] = new ConcurrentHashMap[String, Int]
+  // log recovery thread name -> number of remaining segments map for remainingSegmentsToRecover metric
+  private val numRemainingSegments: ConcurrentMap[String, Int] = new ConcurrentHashMap[String, Int]
```
Could we make these two maps local in loadLogs()?

showuon replied:
Done

junrao reviewed (commented):
@showuon : Thanks for the PR. Made a pass of non-testing files. A few comments below.

showuon committed:
KAFKA-13919: fix race condition and refactor

showuon commented on core/src/main/scala/kafka/log/LogLoader.scala:
```diff
@@ -424,8 +433,13 @@ class LogLoader(
+          removeAndDeleteSegmentsAsync(unflushedIter.toList)
           truncated = true
+          // segment is truncated, so set remaining segments to 0
+          numRemainingSegments.put(threadName, 0)
```
set remaining segments to 0 when segment is truncated due to corruption.

showuon commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -366,25 +392,31 @@ class LogManager(logDirs: Seq[File],
               case e: KafkaStorageException if e.getCause.isInstanceOf[IOException] =>
                 // KafkaStorageException might be thrown, ex: during writing LeaderEpochFileCache
                 // And while converting IOException to KafkaStorageException, we've already handled the exception. So we can ignore it here.
+            } finally {
+              val logLoadDurationMs = time.hiResClockMs() - logLoadStartMs
+              val currentNumLoaded = numLogsLoaded.incrementAndGet()
+              updateNumRemainingLogs(numRemainingLogs, dir.getAbsolutePath, logsToLoad.length - currentNumLoaded)
+              log match {
+                case Some(loadedLog) => info(s"Completed load of $loadedLog with ${loadedLog.numberOfSegments} segments in ${logLoadDurationMs}ms " +
+                  s"($currentNumLoaded/${logsToLoad.length} completed in $logDirAbsolutePath)")
+                case None => info(s"Error while loading logs in $logDir in ${logLoadDurationMs}ms ($currentNumLoaded/${logsToLoad.length} completed in $logDirAbsolutePath)")
+              }
```
Move `currentNumLoaded` and log logic into `finally` block, so that we can make sure we also count for "failed" cases in `currentNumLoaded` result and log.

showuon commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -317,6 +338,11 @@ class LogManager(logDirs: Seq[File],
     val jobs = ArrayBuffer.empty[Seq[Future[_]]]
     var numTotalLogs = 0
+    val curNumRecoveryThreadsPerDataDir = numRecoveryThreadsPerDataDir
+    // log dir path -> number of Remaining logs map for remainingLogsToRecover metric
+    val numRemainingLogs: ConcurrentMap[String, Int] = new ConcurrentHashMap[String, Int]
+    // log recovery thread name -> number of remaining segments map for remainingSegmentsToRecover metric
+    val numRemainingSegments: ConcurrentMap[String, Int] = new ConcurrentHashMap[String, Int]
```
Change to local variables in `loadLogs`

junrao commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -307,6 +309,25 @@ class LogManager(logDirs: Seq[File],
     log
   }
 
+  import java.util.concurrent.ThreadFactory
```
Should we move the import to the beginning?

junrao commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -366,25 +392,31 @@ class LogManager(logDirs: Seq[File],
+            } finally {
+              val logLoadDurationMs = time.hiResClockMs() - logLoadStartMs
+              val currentNumLoaded = numLogsLoaded.incrementAndGet()
+              updateNumRemainingLogs(numRemainingLogs, dir.getAbsolutePath, logsToLoad.length - currentNumLoaded)
```
It seems this could still cause the count in numRemainingLogs to go backward temporarily. For example, a job finished by another thread could sneak in between line 413 and 414.

showuon replied:
Good point! I've changed the type in Map, from `Map[String, Int]` to `Map[String, AtomicInteger]`. And we'll decrement one each time update the `numRemainingLogs`, so that we can make sure the metric number will be decremented in correct order. Thanks.

junrao commented on core/src/test/scala/unit/kafka/log/LogManagerTest.scala:
```diff
@@ -638,6 +641,221 @@ class LogManagerTest {
+
+  @Test
+  def testLogRecoveryMetrics(): Unit = {
+    logManager.shutdown()
```
Do we need this since we call this in tearDown()?

showuon replied:
Yes, we need this. In `setUp`, we'll create a default logManager, but that doesn't fit our test. So I have to shutdown it first, and then create another one below. And the shutdown in `tearDown` will shutdown my new created logManger.

junrao commented on core/src/test/scala/unit/kafka/log/LogManagerTest.scala:
```diff
@@ -638,6 +641,221 @@ class LogManagerTest {
+        // pass mock map for verification later
+        numRemainingSegments = mockMap)
+
+    } .when(spyLogManager).loadLog(any[File], any[Boolean], any[Map[TopicPartition, Long]], any[Map[TopicPartition, Long]],
```
remove space before .`when`?

junrao commented on core/src/test/scala/unit/kafka/log/LogManagerTest.scala:
```diff
@@ -638,6 +641,221 @@ class LogManagerTest {
+          .map { case (_, gauge) => gauge }
+          .asInstanceOf[ArrayBuffer[Gauge[Int]]]
+
+    // expected each log dir has 2 metrics for each thread
```
Hmm, each log dir has 1 metric per thread, right?

showuon replied:
Good catch! Updated.

junrao commented on core/src/test/scala/unit/kafka/log/LogManagerTest.scala:
```diff
@@ -638,6 +641,221 @@ class LogManagerTest {
+        log.appendAsLeader(createRecords, leaderEpoch = 0)
+      }
+
+      assertEquals(expectedSegmentsPerLog, log.numberOfSegments)
```
Hmm, how do we enforce the expected number of segments? Should we explicitly call log.roll()?

showuon replied:
> how do we enforce the expected number of segments?

We can make sure the number of segments because we set "segment.byte=1024", and the dummy record size is 72 bytes each. So that we can confirm how many segments to be created. I've updated the test and add comments.

> should we explicitly call log.roll()?

No, I don't think we need `log.roll()` here, because we only need log segments filled with records for recovery. Besides, we don't want to update recovery checkpoint to affect the remaining segments metric results.

junrao commented on core/src/test/scala/unit/kafka/log/LogManagerTest.scala:
```diff
@@ -638,6 +641,221 @@ class LogManagerTest {
+    // intercept loadLog method to pass expected parameter to do log recovery
+    doAnswer { _ =>
+      // simulate the recovery thread is resized during log recovery
+      spyLogManager.resizeRecoveryThreadPool(recoveryThreadsPerDataDir - 1)
```
Is this necessary? The broker is only online after log recovery.

showuon replied:
Ah, good point again! Yes, I confirmed in both ZK mode and Kraft mode, that broker will be online(ready) after log recovery. So it has no chance to change the config during log recovery. Test updated. Thanks.

cc @tombentley

junrao reviewed (commented):
@showuon : Thanks for the updated PR. A few more comments.

showuon committed:
KAFKA-13919: fix race condition and refactor

showuon commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -307,6 +309,27 @@ class LogManager(logDirs: Seq[File],
+   */
+  private[log] def decNumRemainingLogs(numRemainingLogs: ConcurrentMap[String, AtomicInteger], path: String): Int = {
+    require(path != null, "path cannot be null to update remaining logs metric.")
+    numRemainingLogs.get(path).decrementAndGet()
```
Decrement 1 to `AtomicInteger` object when one log loaded, so that we can make sure the metric can reflect the correct number.

tombentley replied:
Is it not possible to use `numRemainingLogs.compute()` and avoid the need for the `AtomicInteger`?

showuon replied:
Good point, Tom! Replaced AtomicInteger with Int, and use compute instead.
```
numRemainingLogs.compute(path, (_, oldVal) => oldVal - 1)
```
I confirmed from the javadoc[1]
> The entire method invocation is performed atomically.

So it is still thread safe. Thank you for the suggestion!

[1] https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#compute-K-java.util.function.BiFunction-

showuon commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -363,28 +390,32 @@ class LogManager(logDirs: Seq[File],
                 // And while converting IOException to KafkaStorageException, we've already handled the exception. So we can ignore it here.
+            } finally {
+              val logLoadDurationMs = time.hiResClockMs() - logLoadStartMs
+              val remainingLogs = decNumRemainingLogs(numRemainingLogs, dir.getAbsolutePath)
+              val currentNumLoaded = logsToLoad.length - remainingLogs
+              log match {
+                case Some(loadedLog) => info(s"Completed load of $loadedLog with ${loadedLog.numberOfSegments} segments in ${logLoadDurationMs}ms " +
+                  s"($currentNumLoaded/${logsToLoad.length} completed in $logDirAbsolutePath)")
+                case None => info(s"Error while loading logs in $logDir in ${logLoadDurationMs}ms ($currentNumLoaded/${logsToLoad.length} completed in $logDirAbsolutePath)")
```
The log output might not be in order, ex: 
... (11/100 completed in /tmp/kafkaLogs)
... (10/100 completed in /tmp/kafkaLogs)
... (12/100 completed in /tmp/kafkaLogs)

but I think that's less important. Otherwise, we need a lock in the `finally` block, which I think it'll affect the log recovery performance. That said, since we can make sure the metric result is in correct order, the log output not in order should be tolerant. WDYT?

One thing to add, the log output might be out of order behavior was already there before my change. FYI

junrao reviewed (approved):
@showuon : Thanks for the latest update. LGTM. I will let you merge the PR after the tests pass.

showuon commented:
@junrao , thanks for your review! 
@tombentley , do you want to have another look?
Thanks.

tombentley commented on core/src/main/scala/kafka/log/LogLoader.scala:
```diff
@@ -64,6 +65,7 @@ object LogLoader extends Logging {
  * @param recoveryPointCheckpoint The checkpoint of the offset at which to begin the recovery
  * @param leaderEpochCache An optional LeaderEpochFileCache instance to be updated during recovery
  * @param producerStateManager The ProducerStateManager instance to be updated during recovery
+ * @param numRemainingSegments The remaining segments to be recovered in this log
```
```suggestion
 * @param numRemainingSegments The remaining segments to be recovered in this log keyed by recovery thread name
```

showuon committed:
Merge branch 'trunk' of https://github.com/apache/kafka into KAFKA-13919

showuon committed:
KAFKA-13919: use concurrentMap.compute to replace AtomicInteger

showuon commented:
@tombentley , I've addressed your comments in this commit: https://github.com/apache/kafka/pull/12347/commits/42cd5ac41eb2cc83548066ea053dad47a0eddd22 . Thank you.

tombentley reviewed (approved):
Thanks @showuon, LGTM

showuon commented:
Failed tests are unrelated and also failed in trunk build:
```
    Build / JDK 8 and Scala 2.12 / kafka.server.DynamicBrokerReconfigurationTest.testKeyStoreAlter()
    Build / JDK 8 and Scala 2.12 / kafka.server.DynamicBrokerReconfigurationTest.testKeyStoreAlter()
```

showuon commented:
Backport to 3.3 branch. cc @jsancio

showuon commented:
@jsancio , sorry, I just found it's been passed the feature freeze date. But I think this PR is only adding metrics in log manager, it should fine to backport to 3.3. But let me know if you disagree that. Thank you.
